internal/pkg/mining/scheduler.go

```diff
@@ -7,108 +7,127 @@ package mining
-			if !base.Defined() { // Don't try to mine on an unset head.
-				outCh <- NewOutput(nil, errors.New("cannot mine on unset (nil) head"))
-				return
-			}
-			if prevWon && prevBase.Equals(base) {
-				// Skip this round, this likely means that the new head has not propagated yet through the system.
-				// TODO: investigate if there is a better way to handle this situation.
-				continue
-			}
-
-			// Determine how many null blocks we should mine with.
-			nullBlkCount = nextNullBlkCount(nullBlkCount, prevBase, base)
-
-			// Mine synchronously! Ignore all new tipsets.
-			prevWon = s.worker.Mine(miningCtx, base, nullBlkCount, outCh)
-			prevBase = base
-		}
+		s.mineLoop(miningCtx, outCh, skipCh, &doneWg)
+		doneWg.Done()
 	}()
+	s.isStarted = true
 
-	// This tear down goroutine waits for all work to be done before closing
-	// channels.  When this goroutine is complete, external code can
-	// consider the scheduler to be done.
-	extDoneWg.Add(1)
-	go func() {
-		defer extDoneWg.Done()
-		doneWg.Wait()
-		close(outCh)
-	}()
-	return outCh, &extDoneWg
+	return outCh, &doneWg
+}
+
+func (s *timingScheduler) mineLoop(miningCtx context.Context, outCh chan Output, skipCh chan bool, doneWg *sync.WaitGroup) {
+	// mineLoop is the main event loop for the timing scheduler.  It waits for
+	// a new epoch to start, polls the heaviest head, includes the correct number
+	// of null blocks and starts a mining job async.
+	//
+	// If the previous epoch's mining job is not finished it is canceled via the context
+	//
+	// The scheduler will skip mining jobs if the skipping flag is set
+	workContext, workCancel := context.WithCancel(miningCtx) // nolint:staticcheck
+	for {
+		s.waitForEpochStart(miningCtx)
+		select { // check for interrupt during waiting
+		case <-miningCtx.Done():
+			return // nolint:govet
+		default:
+		}
+		workCancel() // cancel any late work from last epoch
+
+		// check if we are skipping and don't mine if so
+		select {
+		case skip := <-skipCh:
+			s.skipping = skip
+		default:
+		}
+		if s.skipping { // don't mine if we are skipping
+			continue
+		}
+
+		workContext, workCancel = context.WithCancel(miningCtx) // nolint: govet
+		base, err := s.pollHeadFunc()
+		if err != nil {
+			log.Errorf("error polling head from mining scheduler %s", err)
+		}
+		h, err := base.Height()
+		if err != nil {
+			log.Errorf("error getting height from base", err)
+		}
+		nullBlkCount := s.calcNullBlks(h)
+		doneWg.Add(1)
+		go func(ctx context.Context) {
+			s.worker.Mine(ctx, base, nullBlkCount, outCh)
+			doneWg.Done()
+		}(workContext)
+	}
+}
+
+func (s *timingScheduler) waitForEpochStart(miningCtx context.Context) {
+	// waitForEpochStart blocks until the next epoch begins
+	currEpoch := uint64(s.chainClock.EpochAtTime(s.chainClock.Now()))
+	nextEpoch := currEpoch + 1
+	nextEpochStart := s.chainClock.StartTimeOfEpoch(nextEpoch)
+	nowB4 := s.chainClock.Now()
+	waitDur := nextEpochStart.Sub(nowB4)
+	newEpochCh := s.chainClock.After(waitDur)
```

Comment:
I think you could push this down to the chain clock too, `newEpochCh := s.chainClock.WaitForEpoch(nextEpoch)`

Reply:
I would push the entire method to `chainClock.WaitNextEpoch()`